gateway/gateway_test.go

```diff
@@ -2747,3 +2751,127 @@ func waitForBackendConfigInit(gw *Handle) {
 		2*time.Second,
 	).Should(BeTrue())
 }
+
+func TestLeakyUploader(t *testing.T) {
+	// Setup Minio test container
+	pool, err := dockertest.NewPool("")
+	require.NoError(t, err)
+
+	t.Run("invalid JSON payload should be uploaded to Minio", func(t *testing.T) {
+		minioContainer, err := minio.Setup(pool, t)
+		require.NoError(t, err)
+
+		// create test gateway with leaky uploader enabled
+		gw, cleanupFn := createTestGatewayWithLeakyUploader(t, minioContainer.Endpoint, minioContainer.AccessKeyID, minioContainer.AccessKeySecret)
+		t.Cleanup(cleanupFn)
+
+		// Prepare invalid JSON payload
+		invalidPayload := []byte(`{"invalid": "json`)
+
+		// Call internal batch handler
+		jobs, err := gw.extractJobsFromInternalBatchPayload("batch", invalidPayload)
+
+		// Verify error response
+		require.Error(t, err)
+		require.Equal(t, response.InvalidJSON, err.Error())
+		require.Nil(t, jobs)
+
+		// verify file was uploaded to Minio
+		require.Eventually(t, func() bool {
+			contents, err := minioContainer.Contents(context.Background(), "gw-failed-events")
+			require.NoError(t, err)
+			return len(contents) != 0
+		}, 2*time.Second, time.Second, "Minio bucket not updated within timeout")
+	})
+
+	t.Run("valid JSON payload should be processed normally", func(t *testing.T) {
+		// Prepare valid payload
+		validPayload := []byte(`[
+			{
+				"properties": {
+					"requestType": "track",
+					"routingKey": "test-key",
+					"workspaceId": "workspace1",
+					"sourceId": "source-id-1",
+					"receivedAt": "2023-01-01T00:00:00Z",
+					"requestIP": "127.0.0.1"
+				},
+				"payload": "{\"type\":\"track\",\"event\":\"TestEvent\"}"
+			}
+		]`)
+
+		minioContainer, err := minio.Setup(pool, t)
+		require.NoError(t, err)
+
+		// create test gateway with leaky uploader enabled
+		gw, cleanupFn := createTestGatewayWithLeakyUploader(t, minioContainer.Endpoint, minioContainer.AccessKeyID, minioContainer.AccessKeySecret)
+		t.Cleanup(cleanupFn)
+
+		// Call internal batch handler
+		jobs, err := gw.extractJobsFromInternalBatchPayload("batch", validPayload)
+
+		// Verify successful processing
+		require.NoError(t, err)
+		require.NotNil(t, jobs)
+		require.Len(t, jobs, 1)
+
+		// Verify no files uploaded to Minio
+		// verify file was uploaded to Minio
+		require.Never(t, func() bool {
```

Comment:
This could give a false positive like for example if there a delay is introduced, then you'd be happy after 2s but the file gets eventually uploaded.
Can we rely on something else here? For example, do we emit a metric after the upload so that we could assert that that metric is emitted and only then check the bucket. It would make the test more robust.